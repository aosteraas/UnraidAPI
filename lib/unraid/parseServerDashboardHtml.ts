import { extractValue } from '@lib/scraper';
import { parse } from 'node-html-parser';

interface ServerCoreDetails {
  title: string;
  cpu: string;
  memory: string;
  motherboard: string;
  diskSpace: string;
  cacheSpace: string;
  version: string;
}

export function parseServerDashboardHtml(rawHtml: string): ServerCoreDetails {
  const document = parse(rawHtml, {
    blockTextElements: { script: false, noscript: false },
  });
  // need to remove `/Dashboard`
  const title = document.querySelector('title').innerText;
  const cpuRaw = document.querySelector('tr.cpu_view td[colspan=3]');

  // const cpuCol = cpu.querySelector('td[colspan=3]');
  const newLineIdx = cpuRaw.textContent.indexOf('\n');
  const cpu = cpuRaw.textContent.substr(0, newLineIdx);

  const moboRaw = document.querySelector('tr.mb_view td[colspan=3]');
  const moboNewLineIdx = moboRaw.structuredText.indexOf('\n');
  const motherboard = moboRaw.structuredText.substr(0, moboNewLineIdx);

  const memory = extractValue(rawHtml, 'Memory<br><span>', '<');

  // const mem = next.find((x) => x.querySelector('.icon-ram') !== null);
  const [rawCacheSpace, rawDiskSpace] = document
    .querySelectorAll('a[title="Go to disk settings"]')
    .map((d) => d.nextElementSibling.structuredText);

  const cacheSpace = extractCacheDiskDetails(rawCacheSpace);
  const arraySpace = extractArrayDiskDetails(rawDiskSpace);

  return {
    title,
    cpu,
    motherboard,
    memory,
    diskSpace: rawDiskSpace,
    cacheSpace: rawCacheSpace,
    ...cacheSpace,
    ...arraySpace,
    version: 'todo',
  };
  // console.log(data);
  // console.log(ramElementSibling.parentNode);
  // console.log(ramIcon.querySelectorAll('span'));
  // return '';
}

function extractCacheDiskDetails(details: string) {
  const diskDetails = details.split(' used of ');
  const data = {
    cacheUsedSpace: diskDetails[0],
    cacheTotalSpace: diskDetails[1].substring(0, diskDetails[1].indexOf(' (')),
    cacheFreeSpace: '',
  };

  const totalSizeAndDenomination = data.cacheTotalSpace.split(' ');
  const usedSizeAndDenomination = data.cacheUsedSpace.split(' ');
  const usedNumber = parseInt(usedSizeAndDenomination[0]);
  let totalNumber = parseInt(totalSizeAndDenomination[0]);

  if (usedSizeAndDenomination[1] !== totalSizeAndDenomination[1]) {
    totalNumber *= 1024;
  }

  const freeNumber = totalNumber - usedNumber;
  data.cacheFreeSpace = `${freeNumber} ${usedSizeAndDenomination[1]}`;
  return data;
}

function extractArrayDiskDetails(details: string) {
  const diskDetails = details.split(' used of ');
  const data = {
    arrayUsedSpace: diskDetails[0],
    arrayTotalSpace: diskDetails[1].substring(0, diskDetails[1].indexOf(' (')),
    arrayFreeSpace: '',
  };

  const totalSizeAndDenomination = data.arrayTotalSpace.split(' ');
  const usedSizeAndDenomination = data.arrayUsedSpace.split(' ');
  const usedNumber = parseInt(usedSizeAndDenomination[0]);
  let totalNumber = parseInt(totalSizeAndDenomination[0]);

  if (usedSizeAndDenomination[1] !== totalSizeAndDenomination[1]) {
    totalNumber *= 1024;
  }

  const freeNumber = totalNumber - usedNumber;
  data.arrayFreeSpace = `${freeNumber} ${usedSizeAndDenomination[1]}`;
  return data;
}
