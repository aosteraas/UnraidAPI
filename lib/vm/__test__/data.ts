export const vmHtml = `<tr parent-id="0" class="sortable">
  <td class="vm-name" style="width: 220px; padding: 8px">
    <span class="outer"
      ><span id="vm-1d608caf-a589-395f-121a-c6c70f9abac6" class="hand"
        ><img
          src="/plugins/dynamix.vm.manager/templates/images/ubuntu.png"
          class="img" /></span
      ><span class="inner"
        ><a
          href="#"
          onclick='return toggle_id("name-0")'
          title="click for more VM info"
          >Lubuntu</a
        ><br /><i class="fa fa-square stopped red-text"></i
        ><span class="state">stopped</span></span
      ></span
    >
  </td>
  <td></td>
  <td>
    <a class="vcpu-1d608caf-a589-395f-121a-c6c70f9abac6" style="cursor: pointer"
      >2</a
    >
  </td>
  <td>4096M</td>
  <td title="Current physical size: 20G">1 / 20G</td>
  <td>VNC:auto</td>
  <td>
    <input
      class="autostart"
      type="checkbox"
      name="auto_Lubuntu"
      title="Toggle VM autostart"
      uuid="1d608caf-a589-395f-121a-c6c70f9abac6"
    />
  </td>
</tr>
<tr child-id="0" id="name-0" style="display: none">
  <td colspan="8" style="margin: 0; padding: 0">
    <table class="tablesorter domdisk" id="domdisk_table">
      <thead>
        <tr>
          <th><i class="fa fa-hdd-o"></i> <b>Disk devices</b></th>
          <th>Bus</th>
          <th>Capacity</th>
          <th>Allocation</th>
        </tr>
      </thead>
      <tbody id="domdisk_list">
        <tr>
          <td>/mnt/disks/vms/Lubuntu/vdisk1.img</td>
          <td>VirtIO</td>
          <td title="Click to increase Disk Size">
            <form method="get" action="">
              <input type="hidden" name="subaction" value="disk-resize" /><input
                type="hidden"
                name="uuid"
                value="1d608caf-a589-395f-121a-c6c70f9abac6"
              /><input
                type="hidden"
                name="disk"
                value="/mnt/disks/vms/Lubuntu/vdisk1.img"
              /><input type="hidden" name="oldcap" value="20G" /><span
                class="diskresize"
                style="width: 30px"
                ><span class="text"
                  ><a href="#" onclick="return false">20G</a></span
                ><input
                  class="input"
                  type="text"
                  style="width: 46px"
                  name="cap"
                  value="20G"
                  val="diskresize"
                  hidden
              /></span>
            </form>
          </td>
          <td>19G</td>
        </tr>
        <tr>
          <td>/mnt/user/isos/lubuntu-20.10-desktop-amd64.iso</td>
          <td>SATA</td>
          <td>2G</td>
          <td>2G</td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
<tr parent-id="1" class="sortable">
  <td class="vm-name" style="width: 220px; padding: 8px">
    <span class="outer"
      ><span id="vm-3c15a77b-7917-514e-91e5-3aa32c29f919" class="hand"
        ><img
          src="/plugins/dynamix.vm.manager/templates/images/ubuntu.png"
          class="img" /></span
      ><span class="inner"
        ><a
          href="#"
          onclick='return toggle_id("name-1")'
          title="click for more VM info"
          >Ubuntu</a
        ><br /><i class="fa fa-square stopped red-text"></i
        ><span class="state">stopped</span></span
      ></span
    >
  </td>
  <td></td>
  <td>
    <a class="vcpu-3c15a77b-7917-514e-91e5-3aa32c29f919" style="cursor: pointer"
      >1</a
    >
  </td>
  <td>1024M</td>
  <td title="Current physical size: 2G">1 / 2G</td>
  <td>VNC:auto</td>
  <td>
    <input
      class="autostart"
      type="checkbox"
      name="auto_Ubuntu"
      title="Toggle VM autostart"
      uuid="3c15a77b-7917-514e-91e5-3aa32c29f919"
    />
  </td>
</tr>
<tr child-id="1" id="name-1" style="display: none">
  <td colspan="8" style="margin: 0; padding: 0">
    <table class="tablesorter domdisk" id="domdisk_table">
      <thead>
        <tr>
          <th><i class="fa fa-hdd-o"></i> <b>Disk devices</b></th>
          <th>Bus</th>
          <th>Capacity</th>
          <th>Allocation</th>
        </tr>
      </thead>
      <tbody id="domdisk_list">
        <tr>
          <td>/mnt/disks/vms/Ubuntu/vdisk1.img</td>
          <td>VirtIO</td>
          <td title="Click to increase Disk Size">
            <form method="get" action="">
              <input type="hidden" name="subaction" value="disk-resize" /><input
                type="hidden"
                name="uuid"
                value="3c15a77b-7917-514e-91e5-3aa32c29f919"
              /><input
                type="hidden"
                name="disk"
                value="/mnt/disks/vms/Ubuntu/vdisk1.img"
              /><input type="hidden" name="oldcap" value="2G" /><span
                class="diskresize"
                style="width: 30px"
                ><span class="text"
                  ><a href="#" onclick="return false">2G</a></span
                ><input
                  class="input"
                  type="text"
                  style="width: 46px"
                  name="cap"
                  value="2G"
                  val="diskresize"
                  hidden
              /></span>
            </form>
          </td>
          <td>1M</td>
        </tr>
      </tbody>
    </table>
  </td>
</tr>
addVMContext('Lubuntu','1d608caf-a589-395f-121a-c6c70f9abac6','Ubuntu','shutoff','','');addVMContext('Ubuntu','3c15a77b-7917-514e-91e5-3aa32c29f919','Ubuntu','shutoff','','');var
kvm=[];kvm.push({id:'1d608caf-a589-395f-121a-c6c70f9abac6',state:'shutoff'});kvm.push({id:'3c15a77b-7917-514e-91e5-3aa32c29f919',state:'shutoff'});`;
