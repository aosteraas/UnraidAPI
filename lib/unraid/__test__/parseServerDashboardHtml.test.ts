import { parseServerDashboardHtml } from '../parseServerDashboardHtml';
import { dashboardHtml } from './data';

test('Parses server dashboard html', async () => {
  const res = parseServerDashboardHtml(dashboardHtml);
  expect(res).toMatchSnapshot();
});
