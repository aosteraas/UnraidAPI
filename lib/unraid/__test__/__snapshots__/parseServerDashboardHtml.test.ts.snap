// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parses server dashboard html 1`] = `
Object {
  "arrayFreeSpace": "831 GB",
  "arrayTotalSpace": "1 TB",
  "arrayUsedSpace": "193 GB",
  "cacheFreeSpace": "14 TB",
  "cacheSpace": "12.4 TB used of 26 TB (47.6 %)",
  "cacheTotalSpace": "26 TB",
  "cacheUsedSpace": "12.4 TB",
  "cpu": "Intel® Core™ i5-9600K CPU @ 3.70GHz",
  "diskSpace": "193 GB used of 1 TB (19.3 %)",
  "memory": "16 GiB DDR4 ",
  "motherboard": "ASRock Z390 Pro4",
  "title": "Tower/Dashboard",
  "version": "todo",
}
`;
